### ISSUE TITLE
$ gem install should look for dependencies in the same directory as the dependant gem

### ISSUE THREAD

**Issue by Maniacal** [rubygems/rubygems#448]
If I want to install a.gem and it depends on b.gem and I want to install them from local files(`/tmp/gemstage/a.gem` and `/tmp/gemstage/b.gem`), I should be able to install them as follows:

```
gem install /tmp/gemstage/a.gem --local
```

But this will fail because it won't know where to look for b.gem.  I can cd to `/tmp/gemstage/` and then run:

```
gem install a.gem --local
```

It would be a nice feature if gem install would look for dependencies in the same folder as the gem you're installing. That way you won't have to be concerned with changing directories when scripting these installations.

**Comment by copiousfreetime** [comment]
To summarize this issue:

Currently rubygems will look in the current directory for additional gems to install when you do a gem installation For example:
```
% ls -l
total 2448
-rw-r--r--  1 jeremy  staff  102400 Jan 29 14:18 addressable-2.4.0.gem
-rw-r--r--  1 jeremy  staff  220672 Jan 29 14:18 excon-0.45.4.gem
-rw-r--r--  1 jeremy  staff   31744 Jan 29 14:18 hitimes-1.2.3.gem
-rw-r--r--  1 jeremy  staff   10240 Jan 29 14:18 little-plugger-1.1.4.gem
-rw-r--r--  1 jeremy  staff   88064 Jan 29 14:18 logging-1.8.2.gem
-rw-r--r--  1 jeremy  staff   25088 Jan 29 14:18 multi_json-1.11.2.gem
-rw-r--r--  1 jeremy  staff  228864 Jan 29 14:18 rack-1.6.4.gem
-rw-r--r--  1 jeremy  staff   18432 Jan 29 14:18 rack-protection-1.5.3.gem
-rw-r--r--  1 jeremy  staff  369664 Jan 29 14:18 sinatra-1.4.7.gem
-rw-r--r--  1 jeremy  staff   65536 Jan 29 14:18 stickler-2.4.2.gem
-rw-r--r--  1 jeremy  staff   48640 Jan 29 14:18 tilt-2.0.2.gem
-rw-r--r--  1 jeremy  staff   26112 Jan 29 14:18 trollop-2.1.2.gem
 % gem install --local stickler-2.4.2.gem 
Successfully installed little-plugger-1.1.4
Successfully installed logging-1.8.2
Successfully installed trollop-2.1.2
Successfully installed excon-0.45.4
Successfully installed addressable-2.4.0
Successfully installed rack-1.6.4
Successfully installed rack-protection-1.5.3
Successfully installed tilt-2.0.2
Successfully installed sinatra-1.4.7
Successfully installed stickler-2.4.2
10 gems installed
```
This is *working as designed*

The suggestion is, when doing a gem install of a local gem, and that local gem is NOT in the current working directory, then rubygems should use the directory the local gem file in the same way that the current directory is used in the above example.

In other words the following should work instead of giving and error.
```
% ls -1 gems/
addressable-2.4.0.gem
excon-0.45.4.gem
hitimes-1.2.3.gem
little-plugger-1.1.4.gem
logging-1.8.2.gem
multi_json-1.11.2.gem
rack-1.6.4.gem
rack-protection-1.5.3.gem
sinatra-1.4.7.gem
stickler-2.4.2.gem
tilt-2.0.2.gem
trollop-2.1.2.gem
%  gem install --local gems/stickler-2.4.2.gem
ERROR:  Could not find a valid gem 'logging' (~> 1.8) in any repository
```

I don't think we should automatically determine directories in this case, i would be more towards updating the `-s` option to allow adding directories as `::Gem::Source::Local` instances and add that to the dependency resolver. Either that or add an additional command line option to `install` to allow adding additional local sources.

Feedback welcome.

**Comment by deivid-rodriguez** [comment]
I like this proposal! My case is very related. I have

```
$ ls -1 **/*.gemspec
decidim-accountability/decidim-accountability.gemspec
decidim-admin/decidim-admin.gemspec
decidim-api/decidim-api.gemspec
decidim-assemblies/decidim-assemblies.gemspec
decidim-budgets/decidim-budgets.gemspec
decidim-comments/decidim-comments.gemspec
decidim-core/decidim-core.gemspec
decidim-debates/decidim-debates.gemspec
decidim-dev/decidim-dev.gemspec
decidim.gemspec
decidim-meetings/decidim-meetings.gemspec
decidim-pages/decidim-pages.gemspec
decidim-participatory_processes/decidim-participatory_processes.gemspec
decidim-proposals/decidim-proposals.gemspec
decidim-surveys/decidim-surveys.gemspec
decidim-system/decidim-system.gemspec
decidim-verifications/decidim-verifications.gemspec
```

If I built each gem in its own folder, it'd be nice to have a way to install the gem built in the root folder so that its dependencies (the gems in subfolders) are automatically picked up locally.

**Comment by bronzdoc** [comment]
@Maniacal are you stil interested in this feature?

**Comment by deivid-rodriguez** [comment]
I think a good solution for this issue would be to promote `gem build -C` to a global flag that can be applied to any command. So,  to get this use case working, you'd do `gem -C /tmp/gemstage/ install --local`.

The case I posted is only partially related and only raises the question whether rubygems should starting looking for gemspecs in the current folder _and subfolders_ like bundler does.

**Comment by gustavothecoder** [comment]
Hello @deivid-rodriguez ! So the idea is to add the `-C` flag to the `gem install` command, right? I would like to work on this feature :)

**Comment by gustavothecoder** [comment]
Is the desired behavior something like this?

https://user-images.githubusercontent.com/57065994/208303401-4942fd95-bcd6-4fd8-8c8e-7aa3e2c03bf0.mp4

**Comment by deivid-rodriguez** [comment]
Yes @gustavothecoder, that's what I meant.

I think ideally though `-C` should be a global option, namely, it applies to `gem` itself, not to any command in particular, so the command line I'm thinking about would be `gem -C ./gems install net-scp-4.0.0.gem --local`. Basically, I taking inspiration from how the `-C` flag to `git` works. Do you think it'd be feasible to implement that?

**Comment by gustavothecoder** [comment]
I do @deivid-rodriguez. But, looks like this solution was not accepted before: https://github.com/rubygems/rubygems/issues/4058. It's OK to follow this approach despite what was discussed before?

**Comment by deivid-rodriguez** [comment]
You're right @gustavothecoder. But I gave it a second thought again now and I still have the same opinion I had originally: I think it's best to mimic other `-C` flags people may already be familiar with like `git -C` or `ruby -C`.

**Comment by gustavothecoder** [comment]
I understand, and I agree with you. Let's follow the global `-C` approach.
